### PR TITLE
Reader function refactor

### PR DESCRIPTION
**Pull request by acquamarin**

### CODE REVIEW

**Review by ray6080 — COMMENTED** [review]
Can we trigger a reload on the benchmark dataset? Just want to set a routine that non-trivial changes to COPY should always run on large dataset.

File: src/include/processor/operator/persistent/reader.h [hunk 1]
```diff
@@ -36,13 +36,14 @@ class Reader : public PhysicalOperator {
...
-    void getNextNodeGroupInParallel(std::shared_ptr<arrow::Table>& table);
+    void getNextNodeGroupInSerial();
+    void getNextNodeGroupInParallel();
 
 private:
     ReaderInfo readerInfo;
     std::shared_ptr<storage::ReaderSharedState> sharedState;
-    std::vector<std::shared_ptr<arrow::RecordBatch>> leftRecordBatches;
+    std::vector<common::ValueVector*> vectorsToRead;
+    std::vector<arrow::ArrayVector> leftArrays;
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/include/processor/operator/persistent/reader.h:46]: As we discussed, let's still store leftArrays, but change vectorsToRead to dataChunk.

File: src/include/processor/operator/persistent/reader.h [hunk 2]
```diff
@@ -36,13 +36,14 @@ class Reader : public PhysicalOperator {
...
-    void getNextNodeGroupInSerial(std::shared_ptr<arrow::Table>& table);
-    void getNextNodeGroupInParallel(std::shared_ptr<arrow::Table>& table);
+    void getNextNodeGroupInSerial();
+    void getNextNodeGroupInParallel();
 
 private:
     ReaderInfo readerInfo;
     std::shared_ptr<storage::ReaderSharedState> sharedState;
-    std::vector<std::shared_ptr<arrow::RecordBatch>> leftRecordBatches;
+    std::vector<common::ValueVector*> vectorsToRead;
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/include/processor/operator/persistent/reader.h:45]: ditto, changed to datachunk

File: src/include/storage/copier/reader_state.h [hunk 3]
```diff
@@ -84,8 +82,8 @@ using init_reader_data_func_t = std::function<std::unique_ptr<ReaderFunctionData
 using count_blocks_func_t =
     std::function<std::vector<FileBlocksInfo>(std::vector<std::string>& paths,
         common::CSVReaderConfig csvReaderConfig, catalog::TableSchema* tableSchema)>;
-using read_rows_func_t = std::function<arrow::RecordBatchVector(
-    const ReaderFunctionData& functionData, common::block_idx_t blockIdx)>;
+using read_rows_func_t = std::function<void(const ReaderFunctionData& functionData,
+    common::block_idx_t blockIdx, std::vector<common::ValueVector*>)>;
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/include/storage/copier/reader_state.h:86]: Changed to datachunk*

File: src/storage/copier/reader_state.cpp [hunk 4]
```diff
@@ -236,6 +240,55 @@ std::unique_ptr<ReaderMorsel> ReaderSharedState::getParallelMorsel() {
     return std::make_unique<ReaderMorsel>();
 }
 
+void ReaderSharedState::appendToArrayVectors(std::vector<arrow::ArrayVector>& arrayVectors,
+    std::vector<ValueVector*>& vectorsToAppend, uint64_t& numRows) {
+    if (vectorsToAppend[0]->dataType.getPhysicalType() == common::PhysicalTypeID::ARROW_COLUMN) {
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/storage/copier/reader_state.cpp:245]: removed if

File: src/include/storage/copier/reader_state.h [hunk 5]
```diff
@@ -148,18 +146,24 @@ class ReaderSharedState {
...
 
-    std::unique_ptr<ReaderMorsel> getSerialMorsel();
+    std::unique_ptr<ReaderMorsel> getSerialMorsel(std::vector<common::ValueVector*> vectorsToRead);
     std::unique_ptr<ReaderMorsel> getParallelMorsel();
 
     inline void lock() { mtx.lock(); }
     inline void unlock() { mtx.unlock(); }
     inline common::row_idx_t& getNumRowsRef() { return std::ref(numRows); }
 
+    static void appendToArrayVectors(std::vector<arrow::ArrayVector>& arrayVectors,
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/include/storage/copier/reader_state.h:156]: No, `appendToArrayVectors` means append data from `valueVector` to `vector<arrow::ArrayVector>`
`appendArrowArraysToVectors` means append data from `ArrayVector` to arrow vector

**Review by ray6080 — APPROVED** [review]

**Pull request merged**